query-planner-js/src/FieldSet.ts

```diff
@@ -7,16 +7,14 @@ import {
   Kind,
   SelectionNode,
   SelectionSetNode,
-  GraphQLObjectType,
   DirectiveNode,
 } from 'graphql';
 import { getResponseName } from './utilities/graphql';
 import { partition, groupBy } from './utilities/array';
+import { Scope } from './Scope';
 
-export interface Field<
-  TParent extends GraphQLCompositeType = GraphQLCompositeType
-> {
-  scope: Scope<TParent>;
+export interface Field {
+  scope: Scope;
```

Comment:
What is the reason for removing the type argument here? From what I remember, there were places where it was useful to guarantee that the scope of a `Field` was an object type.

Reply:
There was a single place we where using `Field<GraphQLObjectType>`, and this was for the argument of the `groupForField` method which is an argument to `splitFields`. Except that this was actually wrong: there is a genuine path (namely, [this call](https://github.com/apollographql/federation/blob/40ad8f545a45a82ad6d34b57dd92971ac223f360/query-planner-js/src/buildQueryPlan.ts#L639)) where the field passed is *not* a `GraphQLObjectType` (notice that the field is force-casted to `Field<GraphQLObjectType>` but that we can absolutely reach this point with, say, an interface). The only reason this didn't break at runtime was that within the  `groupForField` method, the only part that truly cared that the field type was an object type was this call to `getOwningService`: https://github.com/apollographql/federation/blob/40ad8f545a45a82ad6d34b57dd92971ac223f360/query-planner-js/src/buildQueryPlan.ts#L454
And the branch calling that method is only taken if the type is indeed an object type. And taking a step back, the fact that the `groupForField` implementation has this condition:https://github.com/apollographql/federation/blob/40ad8f545a45a82ad6d34b57dd92971ac223f360/query-planner-js/src/buildQueryPlan.ts#L445
makes it clear that the method was never expected to only be applied to case were `parentType` was an object type.

Anyway, all this to say that the one use of `Field<GraphQLObjectType>` we have is incorrect and in need of change. So I figured, if no places remains where the type argument gives us anything, is it worth bothering? And having the type argument meant having it on `Scope`, but in `Scope` that type argument only applies to the most immediate `parentType`, not the other ones in the chain, and that felt a bit "unfair" to me.

Btw, I had meant to add a note on that PR about this change since it's something I happened to run into but wasn't certainly obvious. Sorry for forgetting.

Reply:
That makes sense. I think that forced cast is unsound code that we should remove (and are in the process of through #702), but if it doesn't currently buy us much and we're unsure of the semantics for a type argument on `Scope` I'm fine with removing it. It would be nice to get more type safety back later, but we can tackle that separately.